File: hal/src/main/java/edu/wpi/first/wpilibj/sim/PDPSim.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj.sim;

import edu.wpi.first.hal.sim.mockdata.PDPDataJNI;

public class PDPSim {
  private int m_index;

  public PDPSim(int index) {
    m_index = index;
  }

  public CallbackStore registerInitializedCallback(NotifyCallback callback, boolean initialNotify) {
    int uid = PDPDataJNI.registerInitializedCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, PDPDataJNI::cancelInitializedCallback);
  }
  public boolean getInitialized() {
    return PDPDataJNI.getInitialized(m_index);
  }
  public void setInitialized(boolean initialized) {
    PDPDataJNI.setInitialized(m_index, initialized);
  }

  public CallbackStore registerTemperatureCallback(NotifyCallback callback, boolean initialNotify) {
    int uid = PDPDataJNI.registerTemperatureCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, PDPDataJNI::cancelTemperatureCallback);
  }
  public double getTemperature() {
    return PDPDataJNI.getTemperature(m_index);
  }
  public void setTemperature(double temperature) {
    PDPDataJNI.setTemperature(m_index, temperature);
  }

  public CallbackStore registerVoltageCallback(NotifyCallback callback, boolean initialNotify) {
    int uid = PDPDataJNI.registerVoltageCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, PDPDataJNI::cancelVoltageCallback);
  }
  public double getVoltage() {
    return PDPDataJNI.getVoltage(m_index);
  }
  public void setVoltage(double voltage) {
    PDPDataJNI.setVoltage(m_index, voltage);
  }

  public CallbackStore registerCurrentCallback(int channel, NotifyCallback callback, boolean initialNotify) {
    int uid = PDPDataJNI.registerCurrentCallback(m_index, channel, callback, initialNotify);
    return new CallbackStore(m_index, channel, uid, PDPDataJNI::cancelCurrentCallback);
  }
  public double getCurrent(int channel) {
    return PDPDataJNI.getCurrent(channel, m_index);
  }
  public void setCurrent(int channel, double current) {
    PDPDataJNI.setCurrent(m_index, channel, current);
  }

  public void resetData() {
    PDPDataJNI.resetData(m_index);
  }
}

File: wpilibc/src/main/native/cpp/simulation/PDPSim.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018-2020 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "frc/simulation/PDPSim.h"

#include <memory>
#include <utility>

#include <hal/simulation/PDPData.h>

#include "frc/PowerDistributionPanel.h"

using namespace frc;
using namespace frc::sim;

PDPSim::PDPSim(int module) : m_index{module} {}

PDPSim::PDPSim(const PowerDistributionPanel& pdp) : m_index{pdp.GetModule()} {}

std::unique_ptr<CallbackStore> PDPSim::RegisterInitializedCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelPDPInitializedCallback);
  store->SetUid(HALSIM_RegisterPDPInitializedCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool PDPSim::GetInitialized() const {
  return HALSIM_GetPDPInitialized(m_index);
}

void PDPSim::SetInitialized(bool initialized) {
  HALSIM_SetPDPInitialized(m_index, initialized);
}

std::unique_ptr<CallbackStore> PDPSim::RegisterTemperatureCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelPDPTemperatureCallback);
  store->SetUid(HALSIM_RegisterPDPTemperatureCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

double PDPSim::GetTemperature() const {
  return HALSIM_GetPDPTemperature(m_index);
}

void PDPSim::SetTemperature(double temperature) {
  HALSIM_SetPDPTemperature(m_index, temperature);
}

std::unique_ptr<CallbackStore> PDPSim::RegisterVoltageCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelPDPVoltageCallback);
  store->SetUid(HALSIM_RegisterPDPVoltageCallback(m_index, &CallbackStoreThunk,
                                                  store.get(), initialNotify));
  return store;
}

double PDPSim::GetVoltage() const { return HALSIM_GetPDPVoltage(m_index); }

void PDPSim::SetVoltage(double voltage) {
  HALSIM_SetPDPVoltage(m_index, voltage);
}

std::unique_ptr<CallbackStore> PDPSim::RegisterCurrentCallback(
    int channel, NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, channel, -1, callback, &HALSIM_CancelPDPCurrentCallback);
  store->SetUid(HALSIM_RegisterPDPCurrentCallback(
      m_index, channel, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

double PDPSim::GetCurrent(int channel) const {
  return HALSIM_GetPDPCurrent(m_index, channel);
}

void PDPSim::SetCurrent(int channel, double current) {
  HALSIM_SetPDPCurrent(m_index, channel, current);
}

void PDPSim::GetAllCurrents(double* currents) const {
  HALSIM_GetPDPAllCurrents(m_index, currents);
}

void PDPSim::SetAllCurrents(const double* currents) {
  HALSIM_SetPDPAllCurrents(m_index, currents);
}

void PDPSim::ResetData() { HALSIM_ResetPDPData(m_index); }
